### ISSUE TITLE
Clarification request: GSM8K Table 15 C2C evaluation details

### ISSUE THREAD

**Issue by giedlinlofty118-blip** [thu-nics/C2C#13]
Hi, thanks for the great work. I am trying to reproduce Appendix A.5 / Table 15 on GSM8K, especially the C2C row with 62.55 accuracy.

Could you clarify the exact setup for this row? In particular:
- sharer model names
- which C2C fuser checkpoint was used
- whether the sharer uses the interpreter prompt and receiver uses the solver prompt
- whether any interpreter text is generated for the C2C row, or only for T-C2C
- whether the metric is GSM8K exact-match/final-number accuracy or the Cognify LLM-judge score
- dataset split/sample size/seed
- decoding settings and chat template settings

I tried reproducing with the public code/checkpoints and Cognify math prompts, but my exact-match result is lower, so I may be using a different evaluation protocol. Is there an official script/config for Table 15?

Thanks!

**Comment by minzh23** [comment]
Hi @giedlinlofty118-blip, thanks for your interest in our work.

For the GSM8K C2C result reported in Table 15 (62.55), the setup is:

- Sharer model: Qwen3-4B
- C2C fuser checkpoint:
  https://huggingface.co/nics-efc/C2C_Fuser/tree/main/qwen3_0.6b%2Bqwen3_4b_Fuser

For prompting, the sharer uses the interpreter prompt and the receiver uses the solver prompt, as you described.

Regarding interpreter generation:
- For C2C, no interpreter text is generated. The sharer communicates through KV-cache only.
- The interpreter text generation is only used in T-C2C.

For evaluation, we use the official evaluation script in this repository:
https://github.com/thu-nics/C2C/blob/main/script/evaluation/unified_evaluator.py

Most evaluation details (metric computation, decoding pipeline, etc.) can be found in the implementation of this script.

Hope this helps. If you run into any further issues during reproduction, please feel free to follow up here, and we'll be happy to help.

**Comment by giedlinlofty118-blip** [comment]
Thanks for the clarification. I reran the full GSM8K test split with the public C2C code/checkpoint following your setup:

- receiver/base model: Qwen3-0.6B
- sharer model: Qwen3-4B
- fuser checkpoint: `nics-efc/C2C_Fuser/qwen3_0.6b+qwen3_4b_Fuser`
- sharer prompt: Cognify math interpreter prompt
- receiver prompt: Cognify math solver prompt
- C2C only uses the sharer KV-cache; no interpreter text is generated
- official `script/evaluation/unified_evaluator.py`
- greedy decoding: `do_sample: false`, `max_new_tokens: 256`
- full GSM8K test split: 1319 samples

However, the result is much lower than Table 15:

- full GSM8K C2C: 28.58%
- split accuracies: 31.61%, 32.73%, 26.06%, 23.94%

For reference, on the first 300 GSM8K samples, I got:

- receiver-only Qwen3-0.6B: 43.67%
- C2C: 33.00%

So the gap seems unlikely to be caused only by dataset size.

Could you please share the exact config or command used for the Table 15 GSM8K C2C result? In particular, I would like to confirm:

1. The exact repository commit used for Table 15.
2. The exact YAML/config file, if available.
3. The full decoding settings: `max_new_tokens`, `do_sample`, stop/eos settings, etc.
4. Whether Qwen3 chat template uses `enable_thinking=False`.
5. The exact receiver user prompt for C2C when no interpreter text is generated. Is `math model:` left empty, omitted, or filled with any placeholder?
6. How `kv_cache_index` and position ids are constructed when the sharer interpreter prompt and receiver solver prompt have different token lengths.
7. Whether Table 15 used the current public `unified_evaluator.py` path directly, or an internal/modified evaluator.
8. Whether the metric is the exact final-number accuracy computed by `unified_evaluator.py`, not a Cognify LLM-judge score.

The main ambiguity seems to be the KV-cache alignment/token-position construction for different sharer/receiver prompts. Any official config or minimal reproduction command would be very helpful.

**Comment by minzh23** [comment]
Hi @giedlinlofty118-blip, thanks for the detailed follow-up. I rechecked the experiment settings for the GSM8K Table 15 run, and I need to correct one important point in my previous reply.

The main source of confusion is the fuser checkpoint. The **62.55** GSM8K C2C result in Table 15 was obtained with a **GSM8K-specific C2C fuser** trained for the `Qwen3-0.6B` receiver and `Qwen3-4B` sharer. It was not obtained with the public generic `qwen3_0.6b+qwen3_4b_Fuser` checkpoint.

## Exact Setup for the 62.55 C2C Row

- Receiver/base model: `Qwen/Qwen3-0.6B`
- Sharer model: `Qwen/Qwen3-4B`
- Fuser: a GSM8K-specific C2C fuser trained from Qwen3-4B generated GSM8K training data
- Evaluation dataset: full `openai/gsm8k` test split

So the key correction is:

**Table 15 used a task-specific GSM8K fuser, not the public general-purpose Qwen3-0.6B + Qwen3-4B fuser.**

Below are the settings used for that run.

## 1. Teacher Data Generation for the GSM8K Fuser

The GSM8K-specific fuser was trained on GSM8K train examples with responses generated by `Qwen/Qwen3-4B`.

```yaml
dataset_path: openai/gsm8k
split: train
model_path: Qwen/Qwen3-4B
max_new_tokens: 1024
temperature: 0.7
top_p: 0.8
top_k: 20
```

The prompt used to generate the teacher data was the same solver-style GSM8K prompt:

```text
You are a math solver. Given a math problem, your task is to compute the solution and return the final answer. Be concise and clear in your response.

{question}
```

## 2. GSM8K-specific C2C Fuser Training

The receiver and sharer LMs were frozen; only the C2C projector parameters were trained.

```yaml
model:
  base_model: Qwen/Qwen3-0.6B
  teacher_model: Qwen/Qwen3-4B
  include_response: false
  is_do_alignment: false
  alignment_strategy: first
  mapping: last_aligned
  projector:
    type: C2CProjector
    params:
      hidden_dim: 1024
      intermediate_dim: 1024
      num_layers: 3
      dropout: 0.1
      initial_temperature: 1.0
      final_temperature: 0.001
      anneal_steps: 116

training:
  learning_rate: 0.0001
  weight_decay: 0.01
  num_epochs: 1
  max_length: 1024
  gradient_accumulation_steps: 1
  per_device_train_batch_size: 8
  num_processes: 8
  freeze: [teacher, base]
  seed: 42

data:
  type: LLMGeneratedChatDataset
  train_ratio: 0.99
  kwargs:
    split: train
    max_word_count: 1024
    num_samples: null
    data_path: <qwen3_4b_generated_gsm8k_train_dataset>
```

The training command was the standard C2C fuser training command:

```bash
torchrun --nproc_per_node=8 script/train/SFT_train.py \
  --config <gsm8k_specific_fuser_training_config>
```

## 3. Evaluation Setting for the 62.55 C2C Row

The evaluation used the full GSM8K test split and the GSM8K-specific fuser checkpoint from the training above.

```yaml
eval:
  dataset: gsm8k
  answer_method: generate
  sample_interval: 1
  use_template: true

model:
  model_name: Rosetta
  generation_config:
    do_sample: false
    max_new_tokens: 1024
  rosetta_config:
    base_model: Qwen/Qwen3-0.6B
    teacher_model: Qwen/Qwen3-4B
    checkpoints_dir: <gsm8k_specific_qwen3_0.6b_qwen3_4b_fuser>/final
    include_response: false
    is_do_alignment: false
    alignment_strategy: first

chat_template:
  add_generation_prompt: true
  enable_thinking: false

prompt:
  receiver_user_prompt: |
    You are a math solver. Given a math problem, your task is to compute the solution and return the final answer. Be concise and clear in your response.

    {question}
  generated_interpreter_text: none
```

With `dataset: gsm8k`, the evaluator uses the built-in GSM8K configuration:

- dataset: `openai/gsm8k`
- split: `test`
- sample size: full test split, 1319 examples when `sample_interval: 1` and no `limit` is set
- metric: rule-based GSM8K final-answer equivalence using `GSM8KEvaluator.rule_judge`

The ground-truth answer is parsed from the GSM8K `answer` field after the `####` marker, and the generated answer is extracted and verified by the repository's math evaluator.

The evaluation command was:

```bash
python script/evaluation/unified_evaluator.py \
  --config <gsm8k_c2c_eval_config>
```

Sorry again for the confusion in my earlier reply. The important correction is that the Table 15 62.55 C2C number depends on the GSM8K-specific fuser checkpoint.

**Comment by fuvty** [comment]
Thank you, @minzh23, for the detailed response, and thanks to @giedlinlofty118-blip for the discussion thread. If the questions have been resolved, we can close the issue for now. If there are any remaining questions, we’re happy to address them.
